script/on_tick.lua

```diff
@@ -1,80 +1,90 @@
 local math2d = require('math2d')
-
-local function detect_train(prop,range)
-	temp = {x=-1,y=-range}
-	if prop["arrived"].uarea.x == 0 then
-		left_top = temp
+local constants = require('constants')
+
+local function detect_train(train,type)
+	if type == "Train" then
+		if train.exit_uarea.x == 0 then
+			left_top = {x=-1,y=-10}
+		else
+			left_top = {x=-8,y=-1}
+		end
 	else
-		left_top = {x=temp.y,y=temp.x}
+		if train.exit_uarea.x == 0 then
+			left_top = {x=-1,y=-4.5}
+		else
+			left_top = {x=-3,y=-1}
+		end
 	end
+
 	right_down = math2d.position.multiply_scalar(left_top,-1)
 
-	left_top = math2d.position.add(prop["arrived"].site,left_top)
-	right_down = math2d.position.add(prop["arrived"].site,right_down)
+	left_top = math2d.position.add(train.destination.tunnel.position,left_top)
+	right_down = math2d.position.add(train.destination.tunnel.position,right_down)
 
-	entities = prop.guideCar.surface.find_entities({left_top,right_down})
+	entities = train.ghostCar.surface.find_entities({left_top,right_down})
 	
 	for index, value in ipairs(entities) do
-		if	"locomotive" == value.name or "cargo-wagon" == value.name or "fluid-wagon" == value.name then
+		if	"locomotive" == value.name or "cargo-wagon" == value.name or "fluid-wagon" == value.name then --maybe add car, tank and player
 			return true
 		end
 	end
 
 	return false
 end
 
-local function load_train(prop)
-	NewTrain = prop.guideCar.surface.create_entity({
-			name = prop.name,
-			position = prop["arrived"].site,
-			orientation = prop.orientation,
-			force = prop.guideCar.force,
+local function load_train(train)
+	train.TempTrain2.destroy()
+	NewTrain = train.ghostCar.surface.create_entity({
+			name = train.name,
+			position = train.exit_position,
+			orientation = train.orientation,
+			force = train.ghostCar.force,
 			raise_built = true
 			})
-	if NewTrain then
+	if NewTrain ~= nil then
```

Comment:
Whats the difference between `if A` then and `if A ~= nil` then?
We have similar behaviors in Javascript but JS assumes empty string / 0 as false so in JS `if A` and `if A != null` is little different, but in lua using `if A` seems just fine..